src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -729,3 +776,76 @@ object OutputType extends FgBioEnum[OutputType] {
   case object Bam extends OutputType { val producesBam: Boolean = true; val producesFastq: Boolean = false; }
   case object BamAndFastq extends OutputType { val producesBam: Boolean = true; val producesFastq: Boolean = true; }
 }
+
+/** A little class to store read-level information from the comment in a FASTQ read name that is in Illumina
+  * standard format:
+  *
+  * `<read>:<is filtered>:<control number>:<barcode-sequence>`
+  *
+  * This is typically inferred from the first comment in the input FASTQ header.
+  *
+  * @param readNumber the read number
+  * @param passQc true if the read passes quality filters, false if it fails
+  * @param internalControl true if the read is an internal control, false otherwise
+  * @param sampleInfo additional sample information, such as sample barcode (bcl2fastq/BCL convert) or sample number (BaseSpace)
+  * @param rest any additional information beyond the comment.
+  */
+case class ReadInfo(readNumber: Int, passQc: Boolean, internalControl: Boolean, sampleInfo: String, rest: Seq[String]) {
+  override def toString: String = {
+    val leading = f"$readNumber:${if (passQc) "Y" else "N"}:${if (internalControl) 1 else 0}:$sampleInfo"
+    if (rest.isEmpty) leading else leading + " " + rest.mkString(" ")
+  }
+}
+
+object ReadInfo {
+  /** Builds the [[ReadInfo]] by parsing a [[FastqRecord]]. */
+  def apply(rec: FastqRecord): ReadInfo = this(rec.name, rec.comment)
+
+  /** Builds the [[ReadInfo]] by parsing a [[DemuxRecord]]. */
+  def apply(rec: DemuxRecord): ReadInfo = this(rec.name, rec.comment)
+
+  /** Builds the [[ReadInfo]] by parsing a standard input FASTQ. */
+  def apply(name: String, comment: Option[String]): ReadInfo = try {
+    val commentFields = comment.getOrElse {
+      throw new IllegalArgumentException(s"Comment not found.")
+    }.split(' ')
+    val readInfo = commentFields.head
+
+    require(name.count(_ == ':') == 6,
+      s"Expected the read name format 'Instrument:RunID:FlowCellID:Lane:Tile:X:Y', found '${name}'")
+    require(readInfo.count(_ == ':') == 3,
+      s"Expected the comment format 'ReadNum:FilterFlag:0:SampleNumber', found '$readInfo'")
```

Comment:
Also, why not do the `split` on 819 and then just check the length of the resulting seq instead of scanning the string twice?